src/decoder/mod.rs

```diff
@@ -120,20 +130,29 @@ fn read_f128<R: Read + ?Sized>(reader: &mut R) -> DecoderResult<Decimal128> {
 
 fn decode_array<R: Read + ?Sized>(reader: &mut R, utf8_lossy: bool) -> DecoderResult<Array> {
     let mut arr = Array::new();
+    let length = read_i32(reader)?;
 
-    // disregard the length: using Read::take causes infinite type recursion
-    read_i32(reader)?;
+    let mut buf = vec![0u8; (length as usize) - 4];
+    reader.read_exact(&mut buf)?;
 
+    let mut cursor = std::io::Cursor::new(buf);
```

Comment:
This cursor loop seems like it's used in very similar ways a couple times below. Do you think it's worth abstracting in some way?

Reply:
done